### PR TITLE
[Declarative] Resource IDs: Update resource ID best-effort assignment to be consistent for debugging experience

### PR DESCRIPTION
**Pull request by carlosscastro**
The AutoAssignId function was adding an entry with `'id': '<id>'`, but since it was adding a line, the token ranges computed by `SourceRange` were off. 

Here we 

- Recompute the range after assigning id
- Remove double assignment that was happening before
- Improve stack manipulation logic for corner cases (remove bottom of the call stack since it is a duplicate entry always)
- Re-enable cycle detection test

### CODE REVIEW

**Comment by carlosscastro** [comment]
Tom and I chatted, and outlined a solution that is reflected on the latest code.

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Debugging/ResourceSourceContext.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    /// </summary>
+    internal class ResourceSourceContext : SourceContext
+    {
+        /// <summary>
+        /// Gets a mapping between <see cref="JToken" /> and their respective resource ids. 
+        /// </summary>
+        /// <value>
+        /// A mapping between <see cref="JToken" /> and their respective resource ids.
+        /// </value>
+        internal Dictionary<JToken, string> ResourceIdMap { get; } = new Dictionary<JToken, string>(new JTokenEqualityComparer());
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Debugging/ResourceSourceContext.cs:20]: ConcurrentDictionary?

Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Debugging/ResourceSourceContext.cs:20]: >ResourceIdMap [](start = 44, length = 13)

I would just name this 'DefaultIdMap' and put in SourceContext.

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Resources/ResourceExplorer.cs [hunk 2]
```diff
@@ -639,13 +636,13 @@ private void ResourceProvider_Changed(object sender, IEnumerable<Resource> resou
             }
         }
 
-        private void AutoAssignId(Resource resource, JToken jToken)
+        private void AutoAssignId(Resource resource, JToken jToken, SourceContext sourceContext)
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Resources/ResourceExplorer.cs:639]: >AutoAssignId [](start = 21, length = 12)

can we add a unit test to verify that sourceRange for an object matches to what we think it should be?

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Debugging/ResourceSourceContext.cs (hunk 1 above)
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Debugging/ResourceSourceContext.cs:20]: For now I updated the name, we can chat the location in person, would want to brainstorm on pros / cons of each approach.

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Resources/ResourceExplorer.cs (hunk 2 above)
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Resources/ResourceExplorer.cs:639]: I might need to add internal methods and add InternalsVisibleTo for that, since all of this is private.

Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Resources/ResourceExplorer.cs:646]: Added this test plus a few more to cover all combinations

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Debugging/ResourceSourceContext.cs (hunk 1 above)
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Debugging/ResourceSourceContext.cs]: Renamed to DefaultIdMap, left it in the current class after our call.

Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Debugging/ResourceSourceContext.cs]: Left as Dictionary because the process is single threaded in regards to SourceContext usage, and there is one source context per type loading, so multiple concurrent type loadings will have their own respective SourceContext and no shared state.
